ci/release-info-v3.py

```diff
@@ -0,0 +1,100 @@
+import json
+import os
+import re
+import sys
+from datetime import date
+from subprocess import Popen, PIPE
+
+def check_version(ver):
+    # Checks package version for matching with the current tag reference
+    # print(ver[0])
+    if ref is not None and ref != 'refs/tags/v' + str(ver[0]):
+        return 0
+    else:
+        return 1
+
+def print_error(crate, ver, v):
+    # print(crate, ver[0], v)
+    if not v:
+        print(
+            '::error file={path}::version {0} does not match release tag {1}'
+            .format(str(ver[0]), ref, path = str(crate) + '/Cargo.toml')
+        )
+
+def bundle_version(crates):
+    # Reads package versions from workspace manifest file
+    channel = Popen(
+        ['cargo', 'metadata', '--format-version=1', '--no-deps'],
+        stdout=PIPE
+    )
+
+    # parse json data
+    data = json.load(channel.stdout).get('packages')
+
+    # read, map and assign workspace crates versions to bundle package versions
+    for package_id, _ in enumerate(data):
+        if data[package_id]['name'] in crates:
+            crates[data[package_id]['name']].append(data[package_id]['version'])
+    #       print(package_id, data[package_id]['name'], data[package_id]['version'], crates)
+
+    # Checks package versions of the crates bundle for consistency with the given tag reference
+    consistency = list(map(check_version, list(crates.values())))
+    # print(crates, consistency)
+
+    # print errors for packages which versions didn't match tag reference
+    if not all(consistency):
+        list(map(print_error, list(crates.keys()), list(crates.values()), consistency))
```

Comment:
Since you're modifying crates dictionary anyway, the consistency flag could go as a field in each map value. This is a small stylistic matter, though, this works.
Why is the top-level `list` constructor and a `map` needed? `print_error` is used for the side effect not an output value. A for loop would better illustrate what the code does than the functional obfuscation.

Reply:
+1, use a for loop instead

Reply:
Btw the wrapping list for crates./keys/values are useless here, it would work without the extra allocation.

Reply:
> print_error is used for the side effect not an output value.

Nope - print_error exactly print and output all errors in parallel, and this is how maps works

> Why is the top-level list constructor and a map needed?

Without collecting an output of map (errors) in a list data structure in the main consistency check - there would be no output at all for the program, 'cause map should output error messages into container as collector, this list doesn't assigned to some variable, 'cause this isn't used anywhere else, it's the final step of execution, if consistency isn't reached and some of the package versions didn't correspond to tag and version of the whole bundle.

> Since you're modifying crates dictionary anyway, the consistency flag could go as a field in each map value. This is a small stylistic matter, though, this works.

I thought about that - it's better to separate a bit field mask (consistency of package versions) from main data structure with crates in a bundle (with package names and versions), 'cause the crates dictionary may be used in further for other information from metadata, like paths to manifests (this is planned next change), and binding to array element numbers in that way will decrease the flexibility of future use of main data structure (dict in this case in v3 script). That's why it's separated.

> A for loop would better illustrate what the code does than the functional obfuscation.

> +1, use a for loop instead

In first implementation I was used for loops. But because there's a need to iterate on elements from several data structures I was using map(). There are at least two, or even three data structures used for crate names and versions (dict used for this in v3 script - pairs of elements at least are used in one dict, and this has the same time complexity, only container format changed and space complexity little bit decreased), and bit mask flags/fields for consistency check - there always will be nested for loop (in Python manifest - flat is better than nested!) for traversing all elements, check consistency, output versions and package name, and all in one cycle run.
Maps do this in parallel - it takes elements from every iterable/enumerable data structure queue and runs function for this bunch of elements, thus map works and apply function in parallel for all data structures, has linear time complexity, perfect scalability and readaility.
**map(), reduce(), filter(), all(), even fold/foldl/foldr all has a O(n) linear time complexity. One loop nested in another will cause O(n²) quadratic time complexity.**
Even non-nested loop is way more slower than parallel map - even for such little queue of elements it's sensitive for time of output for CI (even milliseconds counts in CI!). For v2 it even more sensitive 'cause script parse in parallel map a massive json string output of `cargo read-manifest` for every package - if it'll be a loop, even non-nested, this will be way less effective.

[https://www.skoumal.com/en/performance-of-built-in-higher-order-functions-map-filter-reduce-and-flatmap-vs-for-in-loop-in-swift/](https://www.skoumal.com/en/performance-of-built-in-higher-order-functions-map-filter-reduce-and-flatmap-vs-for-in-loop-in-swift/)

And there's no any obfuscation - with map(), reduce(), filter(), all() and so on, code is more clear, plain, flat and readable, than with loops and conditions, and readability always counts.

Reply:
Just trying to help here, it does not work as that in most of the cases.